### DISCUSSION TITLE
Unit question

### DISCUSSION THREAD

**Discussion by stabler** [Aalto-Electric-Drives/motulator#104] [Q&A] [answered]
First of all, thanks for the great work on this project!

I've been trying to line up simulation results, and I've been getting unexpected scaling errors on torque and speed. I think I've tracked it down to the flux calculations here:
https://github.com/Aalto-Electric-Drives/motulator/blob/f19e1262a4e11ede2e93c9f2bd15ccce79118869/motulator/model/sm/_drive.py#L86-L89
https://github.com/Aalto-Electric-Drives/motulator/blob/f19e1262a4e11ede2e93c9f2bd15ccce79118869/motulator/model/sm/_drive.py#L79
https://github.com/Aalto-Electric-Drives/motulator/blob/f19e1262a4e11ede2e93c9f2bd15ccce79118869/motulator/model/sm/_drive.py#L113

I believe these calculations (torque and back-emf) are assuming a flux linkage in units of webers, whereas the comment gives units of Vs. If I adjust the flux number by the difference (2*pi) I get results that look roughly consistent for torque and speed (might be off by a small scale factor).

For example, say I have a motor rated at 200 rpm/V (line-to-line) and 1 pole-pair. I calculate a flux in Vs of `psi_f = 60. / (sqrt(3) * 200 * n_p) = 0.173Vs`. When I simulate with that flux value I see speeds (w_M, mechanical) of ~21 rad/s or 200RPM at voltages of ~5.5V (calculated by multiplying `mdl.data.q` by the bus voltage). If instead I use `psi_f = 0.028 Wb` I see speeds of `w_M = ~1130RPM` at ~5V, which is more consistent with the 200 rpm/V motor constant.

Thanks!

**Comment by mhinkkan** [comment, marked as answer]
@stabler, thank you for the comment and trying _motulator_! 

It seems that there is a confusion or misunderstanding regarding the units. The computation of the equivalent circuit parameters for PM machines based on their datasheets or nameplates can be sometimes tricky, since the given data may vary a lot between different motors. 

The selection between volt-second (Vs) or Weber (Wb) as a unit for the flux linkage is a matter of taste. There should be no factor of $2\pi$ in between them, but these units can be used interchangeably. We have selected to systematically use Vs in docstrings, but they could be replaced with Wb with no other changes. 

Here, I try clarifying our choice of Vs as a unit. By its definition, the flux linkage $\psi$ is the electrical-domain quantity, which is related to the induced voltage $e$ through Faraday’s law $e = \mathrm{d}\psi/\mathrm{d}t$. For this reason, we prefer using volt-second (Vs) as a unit for the flux linkage. On the other hand, the magnetic flux $\phi$ is the magnetic-domain quantity. The magnetic flux is not needed anywhere in _motulator_, but in our research and teaching we systematically use Wb as its unit. Furthermore, these conventions are common in academic literature in general.

The permanent-magnet (PM) flux $\psi_\mathrm{f}$ induces the voltage space vector $e_\mathrm{s} = \mathrm{j}\omega_\mathrm{m}\psi_\mathrm{f}$ in the stator terminals, where $\omega_\mathrm{m} = n_\mathrm{p}\omega_\mathrm{M} = n_\mathrm{p} \cdot 2\pi n_\mathrm{rpm}/60$ is the electrical angular speed of the rotor, $\omega_\mathrm{M}$ is the mechanical angular speed of the rotor, $n_\mathrm{rpm}$ is the rotor speed in r/min, and $n_\mathrm{p}$ is the number of pole pairs. 

Notice also that we use peak-value scaling for space vectors in _motulator_. Consequently, the PM flux linkage $\psi_\mathrm{f} = |e_\mathrm{s}|/|\omega_\mathrm{m}|$ can be expressed as

$$\psi_\mathrm{f} = \frac{\sqrt{2/3} E_\mathrm{ll,rms}}{n_\mathrm{p} \cdot 2\pi n_\mathrm{rpm}/60}$$ 

where $E_\mathrm{ll,rms}$ is the magnitude of the induced voltage, expressed as line-to-line rms voltage. If the induced voltage is given in another manner, please change the scaling accordingly. For industrial-scale machines, rated voltages are typically given as line-to-line rms voltages, but this may vary.

It is also worth noticing that the rated voltage and the induced voltage at the rated speed are not exactly same, due to the voltage drops over the stator resistance and the stator inductance as the current flows under the rated-torque condition.

**Reply by stabler** [reply]
Marko,
Thank you for the detailed explanation- I realize now that I could easily have found the correct relationship between volt-second and Weber, but I read an incorrect source and was too caught up in my own dimensional analysis and confirmation bias to double check! Thanks for providing the equation for reference; that is consistent with the equation I had found to work (accounting for RMS vs peak-to-peak).
